rotkehlchen/accounting/cost_basis/base.py

```diff
@@ -136,42 +150,348 @@ def consume_result(self, used_amount: FVal) -> None:
+    _count: int
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._count = 0
+
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        heapq.heappush(self._acquisitions, (self._count, acquisition))
+        self._count += 1
+
+
+class LIFOCostBasisMethod(BaseCostBasisMethod):
+    """Accounting in LIFO (last-in-first-out) method"""
+    _count: int
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._count = 0
+
     def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
-        self._acquisitions.append(acquisition)
+        heapq.heappush(self._acquisitions, (-self._count, acquisition))
+        self._count += 1
 
 
-class LIFOAcquisitionsOrder(BaseAcquisitionsOrder):
-    """Accounting in LIFO (last-in-first-out) order"""
+class HIFOCostBasisMethod(BaseCostBasisMethod):
+    """Accounting in HIFO (highest-in-first-out) method"""
     def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
-        self._acquisitions.appendleft(acquisition)
+        heapq.heappush(self._acquisitions, (-acquisition.amount, acquisition))
+
+
+class AverageCostBasisMethod(BaseCostBasisMethod):
+    """Accounting in Average Cost Base(ACB) method"""
+    # the current cost base of last event(buy/sell) that occurred
+    current_cost_base: FVal
+    # keeps track of the amount of the asset remaining after evert acquisition or spend
+    remaining_amount: FVal
+    _count: int
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._count = 0
+        self.remaining_amount = ZERO
+        self.current_cost_base = ZERO
+        self._last_acquisition_index = -1
+
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        """
+        This method appends an acquisition to the acquisitions list in FIFO order.
+        It also calculates the average cost base of that acquisition with respect to the
+        previous average cost base.
+        
+        The formula used to calculate the average cost base of an acquisition is:
+        [Previous Total ACB] + [Cost of New Shares] + [Transaction Costs]
+        """  # noqa: E501
+        heapq.heappush(self._acquisitions, (self._count, acquisition))
+        self.current_cost_base += (acquisition.rate * acquisition.amount)
+        self.remaining_amount += acquisition.amount
+        self._count += 1
+
+    def consume_result(self, used_amount: FVal) -> None:
+        assert ZERO <= used_amount <= self._acquisitions[0][1].remaining_amount, f'Used amount must be in the interval [0, {self._acquisitions[0][1].remaining_amount}] but it was {used_amount}'  # noqa: E501
+
+        self._acquisitions[0][1].remaining_amount -= used_amount
+        self.remaining_amount -= used_amount
+        if self._acquisitions[0][1].remaining_amount == ZERO:
+            heapq.heappop(self._acquisitions)
+
+    def calculate_spend_cost_basis(
```

Comment:
Already discussed in discord but also leaving a comment here that if final approach will be to use a loop and processing all the acquisitions in `calculate_spend_cost_basis` of average cost basis, then probably there is no need in copy-pasting the method and it can be reused from `BaseCostBasisMethod`